### PR TITLE
perf(semantic): cache byte->UTF-16 column lookups per host line

### LINKED ISSUE #579: perf(semantic): byte→UTF-16 column conversion rescans the line prefix per token
## Found by hot-path audit

src/analysis/semantic/token_collector.rs:392,402,449,492-493 (helper: src/text/position.rs:123-172) — each collected token converts its start/end byte column with an independent scan from the line start. The all-ASCII fast path keeps the common case cheap, but it still scans the byte-prefix per call, and any **non-ASCII** line falls to the per-char decode loop: true O(tokens × line_len) per line.

### When it hurts
Long lines containing CJK/emoji with many tokens (compounds with the O(k²) overlap split in the same scenario — see #577).

### Fix direction
One incremental sweep per line: tokens are sorted, so walk the line once and emit UTF-16 columns for that line's sorted byte offsets instead of restarting from column 0 per token.

### PR DESCRIPTION
**Pull request by atusy**
## Summary
- `collect_host_tokens` converted each token's byte column to UTF-16 independently via `byte_to_utf16_col`, rescanning the line from column 0 every call. The ASCII fast path keeps this cheap; any non-ASCII line falls to a per-char decode loop, so a line with k tokens paid O(line_len) k times over — true O(tokens × line_len) per line.
- Long CJK/emoji lines with many captures compound with the O(k²) overlap split fixed in #577.
- Added `Utf16LineIndex`: ASCII lines keep the zero-allocation fast path; non-ASCII lines build a sorted `(byte, utf16)` boundary table once (O(line_len)) then resolve each query via binary search (O(log line_len)). Reproduces `byte_to_utf16_col`'s exact fallback semantics.
- Wired in via a `Vec<Option<Utf16LineIndex>>` built lazily and shared across every capture on that line. Keyed by the content-local row (not the host-absolute line), so an injection deep in a large host document only allocates up to its own small line count, not the host's; captures arrive in tree-sitter match order, not sorted by column, so a single forward sweep isn't safe in general.

## Test plan
- [x] `Utf16LineIndex` tests mirror every existing `byte_to_utf16_col` test case (ascii, japanese, mixed, mid-character fallback, out-of-bounds)
- [x] New `collect_host_tokens` test with two identifiers straddling a 3-char CJK string literal on one line, pinning that byte/UTF-16 columns diverge correctly and the cache is shared across both lookups
- [x] `cargo test --lib` (2468 tests) all pass
- [x] `cargo clippy --all-targets --all-features -- -D warnings` clean
- [x] Full e2e suite green

Closes #579

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved UTF-16 cursor and token column/length accuracy for non-ASCII text, including Japanese/CJK and mixed-content lines.
  * Fixed incorrect token column/length reporting when multiple tokens appear after CJK literals on the same line.
  * Improved correctness for boundary-adjacent positions (including emoji) and for multiline tokens spanning multiple lines.
* **Tests**
  * Added unit coverage to validate UTF-16 column/length for multiple tokens after CJK prefixes, plus boundary and out-of-range edge cases.

### CODE REVIEW

File: src/analysis/semantic/token_collector.rs [hunk 1]
```diff
@@ -25,6 +27,25 @@ fn is_in_exclusion_range(node: &Node, ranges: &[(usize, usize)]) -> bool {
...
+    cache: &mut HashMap<usize, Utf16LineIndex>,
+    host_line: usize,
+    line_text: &str,
+    byte_col: usize,
+) -> usize {
+    cache
+        .entry(host_line)
+        .or_insert_with(|| Utf16LineIndex::new(line_text))
+        .byte_to_utf16(byte_col)
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 4f2fbfb: introduced cached_line_index returning &Utf16LineIndex, so each token now does one HashMap lookup and calls byte_to_utf16 twice on the reference instead of two separate lookups.

File: src/analysis/semantic/token_collector.rs [hunk 2]
```diff
@@ -389,7 +414,12 @@ pub(super) fn collect_host_tokens(
                 } else {
                     start_pos.column
                 };
-                let start_utf16 = byte_to_utf16_col(host_line_text, byte_offset_in_host);
+                let start_utf16 = cached_utf16_col(
+                    &mut utf16_cache,
+                    host_line,
+                    host_line_text,
+                    byte_offset_in_host,
+                );
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 4f2fbfb: introduced cached_line_index returning &Utf16LineIndex, so each token now does one HashMap lookup and calls byte_to_utf16 twice on the reference instead of two separate lookups.

File: src/analysis/semantic/token_collector.rs [hunk 3]
```diff
@@ -399,7 +429,12 @@ pub(super) fn collect_host_tokens(
                 } else {
                     end_pos.column
                 };
-                let end_utf16 = byte_to_utf16_col(host_line_text, end_byte_offset_in_host);
+                let end_utf16 = cached_utf16_col(
+                    &mut utf16_cache,
+                    host_line,
+                    host_line_text,
+                    end_byte_offset_in_host,
+                );
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 4f2fbfb: introduced cached_line_index returning &Utf16LineIndex, so each token now does one HashMap lookup and calls byte_to_utf16 twice on the reference instead of two separate lookups.

File: src/analysis/semantic/token_collector.rs [hunk 4]
```diff
@@ -429,7 +464,12 @@ pub(super) fn collect_host_tokens(
                     } else {
                         start_pos.column
                     };
-                    let start_utf16 = byte_to_utf16_col(host_start_line_text, start_byte_offset);
+                    let start_utf16 = cached_utf16_col(
+                        &mut utf16_cache,
+                        host_start_line,
+                        host_start_line_text,
+                        start_byte_offset,
+                    );
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 4f2fbfb: introduced cached_line_index returning &Utf16LineIndex, so each token now does one HashMap lookup and calls byte_to_utf16 twice on the reference instead of two separate lookups.

File: src/analysis/semantic/token_collector.rs [hunk 5]
```diff
@@ -446,8 +486,10 @@ pub(super) fn collect_host_tokens(
                             &prefix_widths,
                         );
 
-                        let line_start_utf16 = byte_to_utf16_col(line_text, line_start);
-                        let line_end_utf16 = byte_to_utf16_col(line_text, line_end);
+                        let line_start_utf16 =
+                            cached_utf16_col(&mut utf16_cache, host_row, line_text, line_start);
+                        let line_end_utf16 =
+                            cached_utf16_col(&mut utf16_cache, host_row, line_text, line_end);
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 4f2fbfb: introduced cached_line_index returning &Utf16LineIndex, so each token now does one HashMap lookup and calls byte_to_utf16 twice on the reference instead of two separate lookups.

File: src/analysis/semantic/token_collector.rs [hunk 6]
```diff
@@ -489,8 +531,18 @@ pub(super) fn collect_host_tokens(
...
+                            host_row,
+                            host_line_text,
+                            line_start_byte,
+                        );
+                        let end_utf16 = cached_utf16_col(
+                            &mut utf16_cache,
+                            host_row,
+                            host_line_text,
+                            line_end_byte,
+                        );
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 4f2fbfb: introduced cached_line_index returning &Utf16LineIndex, so each token now does one HashMap lookup and calls byte_to_utf16 twice on the reference instead of two separate lookups.

File: src/text/position.rs [hunk 7]
```diff
@@ -222,6 +272,77 @@ mod tests {
...
+                index.byte_to_utf16(byte_col),
+                byte_to_utf16_col(line, byte_col)
+            );
+        }
+    }
+
+    #[test]
+    fn utf16_line_index_matches_byte_to_utf16_col_out_of_bounds() {
+        // Past the end of the line, on both ASCII and non-ASCII content —
+        // exercises the "no boundary found" fallback that clamps to line_len.
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/text/position.rs]: Fixed in 4f2fbfb: clarified the comment — ASCII lines clamp to byte length (== utf16 offset), non-ASCII lines clamp to the line's UTF-16 width.

File: src/analysis/semantic/token_collector.rs [hunk 8]
```diff
@@ -778,6 +830,50 @@ mod tests {
...
+            &build_line_start_bytes(code),
+            0,
+            0,
+            false,
+            &[],
+            &[],
+            &mut tokens,
+        );
+
+        assert_eq!(tokens.len(), 2);
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs:857]: Fixed in 4f2fbfb: sort tokens by column before asserting on tokens[0]/tokens[1].

File: src/analysis/semantic/token_collector.rs [hunk 9]
```diff
@@ -3,8 +3,10 @@
 //! This module handles the collection of raw tokens from a single document's
 //! highlight query, including multiline token handling and byte-to-UTF16 conversion.
 
+use std::collections::HashMap;
+
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 0aec4504: utf16_cache is now Vec<Option<Utf16LineIndex>>, indexed directly and resized on first use of a line number; HashMap import removed.

File: src/analysis/semantic/token_collector.rs [hunk 10]
```diff
@@ -25,6 +27,26 @@ fn is_in_exclusion_range(node: &Node, ranges: &[(usize, usize)]) -> bool {
...
+/// line_len) lookups.
+fn cached_line_index<'a>(
+    cache: &'a mut HashMap<usize, Utf16LineIndex>,
+    host_line: usize,
+    line_text: &str,
+) -> &'a Utf16LineIndex {
+    cache
+        .entry(host_line)
+        .or_insert_with(|| Utf16LineIndex::new(line_text))
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs:57]: Fixed in 0aec4504: utf16_cache is now Vec<Option<Utf16LineIndex>>, indexed directly and resized on first use of a line number; HashMap import removed.

File: src/analysis/semantic/token_collector.rs [hunk 11]
```diff
@@ -342,6 +364,10 @@ pub(super) fn collect_host_tokens(
     // Split content text into lines for byte offset calculations
     let content_lines: Vec<&str> = text.lines().collect();
 
+    // Lazily-built byte->UTF-16 lookup per host line, shared across every
+    // capture below. See `cached_line_index`.
+    let mut utf16_cache: HashMap<usize, Utf16LineIndex> = HashMap::new();
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 0aec4504: utf16_cache is now Vec<Option<Utf16LineIndex>>, indexed directly and resized on first use of a line number; HashMap import removed.

File: src/analysis/semantic/token_collector.rs [hunk 12]
```diff
@@ -25,6 +25,32 @@ fn is_in_exclusion_range(node: &Node, ranges: &[(usize, usize)]) -> bool {
...
+fn cached_line_index<'a>(
+    cache: &'a mut Vec<Option<Utf16LineIndex>>,
+    host_line: usize,
+    line_text: &str,
+) -> &'a Utf16LineIndex {
+    if host_line >= cache.len() {
+        cache.resize_with(host_line + 1, || None);
+    }
+    cache[host_line].get_or_insert_with(|| Utf16LineIndex::new(line_text))
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 866db6d5: cached_line_index is now keyed by the content-local row (start_pos.row / the multiline loop's row) instead of the absolute host_line, bounding cache size to the content's own line count regardless of where an injection sits in the host document.

File: src/analysis/semantic/token_collector.rs [hunk 13]
```diff
@@ -389,7 +419,8 @@ pub(super) fn collect_host_tokens(
                 } else {
                     start_pos.column
                 };
-                let start_utf16 = byte_to_utf16_col(host_line_text, byte_offset_in_host);
+                let line_index = cached_line_index(&mut utf16_cache, host_line, host_line_text);
+                let start_utf16 = line_index.byte_to_utf16(byte_offset_in_host);
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 866db6d5: cached_line_index is now keyed by the content-local row (start_pos.row / the multiline loop's row) instead of the absolute host_line, bounding cache size to the content's own line count regardless of where an injection sits in the host document.

File: src/analysis/semantic/token_collector.rs [hunk 14]
```diff
@@ -429,7 +460,9 @@ pub(super) fn collect_host_tokens(
                     } else {
                         start_pos.column
                     };
-                    let start_utf16 = byte_to_utf16_col(host_start_line_text, start_byte_offset);
+                    let start_utf16 =
+                        cached_line_index(&mut utf16_cache, host_start_line, host_start_line_text)
+                            .byte_to_utf16(start_byte_offset);
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs:470]: Fixed in 866db6d5: cached_line_index is now keyed by the content-local row (start_pos.row / the multiline loop's row) instead of the absolute host_line, bounding cache size to the content's own line count regardless of where an injection sits in the host document.

File: src/analysis/semantic/token_collector.rs [hunk 15]
```diff
@@ -446,8 +479,9 @@ pub(super) fn collect_host_tokens(
                             &prefix_widths,
                         );
 
-                        let line_start_utf16 = byte_to_utf16_col(line_text, line_start);
-                        let line_end_utf16 = byte_to_utf16_col(line_text, line_end);
+                        let line_index = cached_line_index(&mut utf16_cache, host_row, line_text);
+                        let line_start_utf16 = line_index.byte_to_utf16(line_start);
+                        let line_end_utf16 = line_index.byte_to_utf16(line_end);
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 866db6d5: cached_line_index is now keyed by the content-local row (start_pos.row / the multiline loop's row) instead of the absolute host_line, bounding cache size to the content's own line count regardless of where an injection sits in the host document.

File: src/analysis/semantic/token_collector.rs [hunk 16]
```diff
@@ -489,8 +523,10 @@ pub(super) fn collect_host_tokens(
                             &prefix_widths,
                         );
 
-                        let start_utf16 = byte_to_utf16_col(host_line_text, line_start_byte);
-                        let end_utf16 = byte_to_utf16_col(host_line_text, line_end_byte);
+                        let line_index =
+                            cached_line_index(&mut utf16_cache, host_row, host_line_text);
+                        let start_utf16 = line_index.byte_to_utf16(line_start_byte);
+                        let end_utf16 = line_index.byte_to_utf16(line_end_byte);
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 866db6d5: cached_line_index is now keyed by the content-local row (start_pos.row / the multiline loop's row) instead of the absolute host_line, bounding cache size to the content's own line count regardless of where an injection sits in the host document.

File: src/analysis/semantic/token_collector.rs [hunk 17]
```diff
@@ -25,6 +25,32 @@ fn is_in_exclusion_range(node: &Node, ranges: &[(usize, usize)]) -> bool {
...
+/// Returning the index (rather than resolving one column per call) lets a
+/// caller reuse a single lookup for both the start and end column of one
+/// token, on top of caching by line number turning what would be an
+/// O(line_len) rescan per query into one O(line_len) build plus O(log
+/// line_len) lookups.
+///
+/// `host_line` is a dense index bounded by the host document's line count
+/// (never sparse/huge), so a `Vec<Option<_>>` indexed directly beats a
+/// `HashMap` here: O(1) with no hashing and better cache locality, at the
+/// cost of resizing to cover line numbers as they're first seen.
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 866db6d5: cached_line_index is now keyed by the content-local row (start_pos.row / the multiline loop's row) instead of the absolute host_line, bounding cache size to the content's own line count regardless of where an injection sits in the host document.

File: src/analysis/semantic/token_collector.rs [hunk 18]
```diff
@@ -25,6 +25,32 @@ fn is_in_exclusion_range(node: &Node, ranges: &[(usize, usize)]) -> bool {
...
+fn cached_line_index<'a>(
+    cache: &'a mut Vec<Option<Utf16LineIndex>>,
+    host_line: usize,
+    line_text: &str,
+) -> &'a Utf16LineIndex {
+    if host_line >= cache.len() {
+        cache.resize_with(host_line + 1, || None);
+    }
+    cache[host_line].get_or_insert_with(|| Utf16LineIndex::new(line_text))
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs]: Fixed in 866db6d5: cached_line_index is now keyed by the content-local row (start_pos.row / the multiline loop's row) instead of the absolute host_line, bounding cache size to the content's own line count regardless of where an injection sits in the host document.

File: src/analysis/semantic/token_collector.rs [hunk 19]
```diff
@@ -25,6 +25,37 @@ fn is_in_exclusion_range(node: &Node, ranges: &[(usize, usize)]) -> bool {
...
+fn cached_line_index<'a>(
+    cache: &'a mut Vec<Option<Utf16LineIndex>>,
+    row: usize,
+    line_text: &str,
+) -> &'a Utf16LineIndex {
+    if row >= cache.len() {
+        cache.resize_with(row + 1, || None);
+    }
+    cache[row].get_or_insert_with(|| Utf16LineIndex::new(line_text))
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs:57]: Refuting: `utf16_cache` is call-scoped (allocated fresh at the top of `collect_host_tokens`, dropped when it returns) — neither structure retains memory across requests, so the transient allocation size is the actual cost, not steady-state churn.

For the dominant case — `collect_host_tokens` invoked once per injection layer, where `row` is content-local and bounded by that layer's own (typically small) line count — the Vec is dense and a single flat allocation beats HashMap's per-entry bucket allocation + SipHash lookup cost.

For the top-level host-document call (`content_start_line == 0`), `row` does span the full document, but this is unchanged from before this fix and inherent to any per-line cache over that call regardless of data structure — a HashMap populated for every touched line in a real source file (which has tokens on nearly every non-blank line: keywords, identifiers, punctuation) ends up just as dense. And per `Utf16LineIndex::new` (src/text/position.rs:191-197), ASCII lines — the overwhelming majority — store `boundaries: None`, so each `Vec` slot costs only the `Option` discriminant + one `usize`, not a heap allocation; the 50,000-line adversarial example costs one ~800KB transient Vec, not per-entry churn.

This also reverses the direction of the *previous* round's finding (5 Gemini threads + 1 CodeRabbit thread requesting exactly this Vec+content-local-row design over a `host_line`-keyed HashMap, fixed in 866db6d5f) — keeping the current design per that guidance. Not reworking further; see also src/analysis/semantic/token_collector.rs doc comment on `cached_line_index` explaining the content-local-row invariant.

File: src/analysis/semantic/token_collector.rs [hunk 20]
```diff
@@ -342,6 +373,10 @@ pub(super) fn collect_host_tokens(
     // Split content text into lines for byte offset calculations
     let content_lines: Vec<&str> = text.lines().collect();
 
+    // Lazily-built byte->UTF-16 lookup per host line, shared across every
+    // capture below. See `cached_line_index`.
+    let mut utf16_cache: Vec<Option<Utf16LineIndex>> = Vec::new();
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs:378]: Refuting: `utf16_cache` is call-scoped (allocated fresh at the top of `collect_host_tokens`, dropped when it returns) — neither structure retains memory across requests, so the transient allocation size is the actual cost, not steady-state churn.

For the dominant case — `collect_host_tokens` invoked once per injection layer, where `row` is content-local and bounded by that layer's own (typically small) line count — the Vec is dense and a single flat allocation beats HashMap's per-entry bucket allocation + SipHash lookup cost.

For the top-level host-document call (`content_start_line == 0`), `row` does span the full document, but this is unchanged from before this fix and inherent to any per-line cache over that call regardless of data structure — a HashMap populated for every touched line in a real source file (which has tokens on nearly every non-blank line: keywords, identifiers, punctuation) ends up just as dense. And per `Utf16LineIndex::new` (src/text/position.rs:191-197), ASCII lines — the overwhelming majority — store `boundaries: None`, so each `Vec` slot costs only the `Option` discriminant + one `usize`, not a heap allocation; the 50,000-line adversarial example costs one ~800KB transient Vec, not per-entry churn.

This also reverses the direction of the *previous* round's finding (5 Gemini threads + 1 CodeRabbit thread requesting exactly this Vec+content-local-row design over a `host_line`-keyed HashMap, fixed in 866db6d5f) — keeping the current design per that guidance. Not reworking further; see also src/analysis/semantic/token_collector.rs doc comment on `cached_line_index` explaining the content-local-row invariant.

[the remark this answers is not in the document]
Comment (atusy) [reply src/analysis/semantic/token_collector.rs:378]: Fixed — updated the PR description to say `Vec<Option<Utf16LineIndex>>` keyed by content-local row, matching the implementation.

File: src/text/position.rs [hunk 21]
```diff
@@ -171,6 +171,56 @@ pub fn convert_byte_to_utf16_in_line(line_text: &str, byte_pos: usize) -> Option
...
+impl Utf16LineIndex {
+    pub(crate) fn new(line: &str) -> Self {
+        if line.as_bytes().is_ascii() {
+            return Self {
+                line_len: line.len(),
+                boundaries: None,
+            };
+        }
+
+        let mut boundaries = Vec::with_capacity(line.len() + 1);
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/text/position.rs:199]: Keeping as-is: `line.len() + 1` is a safe single-allocation upper bound (char count ≤ byte count in UTF-8, so it never under-allocates) and, importantly, the CPU-perf-optimal choice for this exact non-ASCII hot path this PR targets — the precise fix (`line.chars().count() + 1`) costs a second full pass over the line before the existing one, and dropping the capacity hint entirely trades the over-allocation for repeated reallocations via amortized doubling. Both alternatives cost more CPU to save a transient, bounded (≤4x) allocation that's freed immediately — `Utf16LineIndex` is scoped to one `collect_host_tokens` call. Not worth an extra pass on the hot path to shrink a short-lived allocation.
